### DISCUSSION TITLE
How to Call Oracle Stored Procedure with Custom Type Out Parameter

### DISCUSSION THREAD

**Discussion by ntubati** [godror/godror#131] [Q&A] [answered]
Hi,

I have a requirement to call a stored oracle procedure which has an out parameter which is a custom type. One of the attributes of that custom type is BLOB. I am not able to get the BLOB contents from that attribute. I am able to call a procedure with the custom type that has non BLOB, CLOB columns and get the values mapped to go struct successfully. But if the oracle custom type has a BLOB attribute it is not working.

Here is my code can someone please help?

// Oracle Custom Type
//----------------------
CREATE OR REPLACE TYPE XX_TEST_OBJ FORCE AS OBJECT
(
FILE_NAME VARCHAR2 (200),
FILE_TYPE VARCHAR2(30),
FILE_CONTENTS BLOB
)

// Oracle Procedure
//-----------------
CREATE OR REPLACE PROCEDURE xxtest_procedure(
p_cco_id IN VARCHAR2,
p_partner_name IN VARCHAR2,
p_tp_ref_num IN VARCHAR2,
p_shipind_docs_o OUT XX_TEST_OBJ,
p_return_code_o OUT VARCHAR2,
p_return_message_o OUT VARCHAR2)
IS

l_shipind_docs_o XX_TEST_OBJ;

BEGIN

l_shipind_docs_o :=
XX_TEST_OBJ (NULL,
NULL,
NULL);

 SELECT file_name,file_type, file_contents
 INTO l_shipind_docs_o.file_name,l_shipind_docs_o.file_type,l_shipind_docs_o.file_contents
 FROM xx_test_table
WHERE file_type = ‘e-Invoice’;

p_shipind_docs_o := l_shipind_docs_o;
p_return_code_o := ‘SUCCESS’;
p_return_message_o := ‘SUCCESS’;

END;
/

//GO Code That I am using
package main

import (
"context"
"database/sql"
"fmt"
"reflect"
"github.com/godror/godror"
_ "github.com/godror/godror"
)

// FileData struct
type FileData struct {
FileName     string
FileType     string
FileContents []byte
}

// SetValues function
func (r *FileData) SetValues(src interface{}) error {
obj, ok := src.(*godror.Object)
if !ok {
    return fmt.Errorf("Cannot scan from type %T", src)
}

fName, err := obj.Get("FILE_NAME")
if err != nil {
    return err
}
r.FileName = string(fName.([]byte))

fType, err := obj.Get("FILE_TYPE")
if err != nil {
    return err
}
r.FileType = string(fType.([]byte))

fContents, err := obj.Get("FILE_CONTENTS")
if err != nil {
    return err
}
r.FileContents = []byte(fContents.([]byte))

return nil
}

func main() {
db, err := sql.Open("godror", "xxeem/*****@ORAPOCDB")
if err != nil {
    fmt.Println(err)
    return
}

defer db.Close()
fmt.Println("Before Prepare Statement")
stmt, err := db.Prepare("BEGIN xxtest_procedure( :p_cco_id,:p_partner_name,:p_tp_ref_num,:p_shipind_docs_o,:p_return_code_o,:p_return_message_o); END;")
fmt.Println("After Prepare", err)
var code, message string
conn, err := db.Conn(context.TODO())
var lFileData FileData
fObjectType, err := godror.GetObjectType(context.TODO(), conn, "XX_TEST_OBJ")
fObject, err := fObjectType.NewObject()

fmt.Println("Before Execute Statement")

_, err = stmt.ExecContext(context.TODO(),
    sql.Named("p_cco_id", "Test"),
    sql.Named("p_partner_name", "Test"),
    sql.Named("p_tp_ref_num", "Test"),
    sql.Named("p_shipind_docs_o", sql.Out{Dest: fObject}),
    sql.Named("p_return_code_o", sql.Out{Dest: &code}),
    sql.Named("p_return_message_o", sql.Out{Dest: &message}),
)

defer stmt.Close()

lFileData.SetValues(fObject)
 
fmt.Println(lFileData)

}

// Few Observations  I have when I am doing trial and error.
Direct Query on Table with Blob column
-- Go Native type used for mapping to Oracle BLOB is [] byte
-- Able to query the oracle table and load blob data into  []byte variable without any issues.
-- Code used in db.Query() and for rows.Next()

Stored Procedure Call with BLOB OUT Variable. No Custom Type
-- godror.LOB column used 
-- mapped the godror.LOB column with the out variable of the procedure
-- io.util.read is used to read data from godror.LOB ( seems like this LOB Variable has the oracle connection details )

Stored Procedure Call with Oracle Table Type Out Variable without any BLOB or CLOB Columns in the table type.
 -- Struct is used 
 -- Object is created with the structure of oracle type
 -- Able to call and retrieve values for all the attributes.
 
Stored Procedure Call with Oracle Table Type Out Variable with BLOB.
 -- Struct is used 
 -- Not Working unable to receive values for the BLOB Column.

**Comment by tgulacsi** [comment]
What is the error message?

You should use the same conn for Prepare/Exec as for GetObjectType though.

**Reply by ntubati** [reply]
I am getting error : getSize: driver: bad connection
I am able to get the other two attributes of the custom type .. but the BLOB Type is not able to


// SetValues function
func (r *FileData) SetValues(src interface{}) error {
obj, ok := src.(*godror.Object)
	if !ok {
		return fmt.Errorf("Cannot scan from type %T", src)
	}
	fName, err := obj.Get("FILE_NAME")
	if err != nil {
		return err
	}
	r.FileName = string(fName.([]byte))

	fType, err := obj.Get("FILE_TYPE")
	if err != nil {
		return err
	}
	r.FileType = string(fType.([]byte))
	var data godror.Data
	err = obj.GetAttribute(&data, "FILE_CONTENTS")
	fmt.Println("GETOBJT Error:", err)
	d, err := ioutil.ReadAll(data.GetLob())
	r.FileContents = d
	fmt.Println(err)
	return nil
}

**Reply by tgulacsi** [reply]
Do you use the same conn for Prepare/Exec as for GetObjectType ???

**Reply by ntubati** [reply]
db, err := sql.Open("godror", "xxeem/****@ORAPOCDB")
defer db.Close()

var code, message string
ctx := context.TODO()
conn, err := db.Conn(ctx)

fObjectType, err := godror.GetObjectType(ctx, conn, "XX_TEST_OBJ")
fObject, err := fObjectType.NewObject()

stmt, err := db.PrepareContext(ctx, "BEGIN xxtest_procedure( :p_cco_id,:p_partner_name,:p_tp_ref_num,:p_shipind_docs_o,:p_return_code_o,:p_return_message_o); END;")

_, err = stmt.ExecContext(ctx,
		sql.Named("p_cco_id", "Test"),
		sql.Named("p_partner_name", "Test"),
		sql.Named("p_tp_ref_num", "Test"),
		sql.Named("p_shipind_docs_o", sql.Out{Dest: fObject}),
		sql.Named("p_return_code_o", sql.Out{Dest: &code}),
		sql.Named("p_return_message_o", sql.Out{Dest: &message}),
	)

**Reply by ntubati** [reply]
Sorry My bad, I am new to golang and still learning..
I realized I wasn't using the same connection object for prepare statement.
I changed it and now I think I am able to extract the blob contents from the returned object type

I was using db.PrepareContext instead I should use conn.PrepareContext
But interestingly thought I was able to get the values from the other two attributes which are VARCHAR2 even when I was using db.PrepareContext.. Not sure what's the logic behind it..

**Reply by tgulacsi** [reply]
Not.much. Technically it's absolutely reasonable, but not too sane.

Everything tha needs the connection (LOB, queue) must use the same connection. An object's scalar fields are retrieved, no connection is needed afterwards, but a LOB is just a reference, so the same connection is needed for retrieving the contents.

**Reply by ntubati** [reply]
Thank You I have a follow-up question on the same .. So does this mean in this scenario to retrieve BLOB Columns there will be multiple round trips to Oracle Database?

I see the below information in godror documentation
*************************************************************************************
LOB as a reader and writer is not the most performant at all. Yes, OCI and ODPI-C provide a way to retrieve this data directly. Effectively, all you need to do is tell ODPI-C that you want a "long string" or "long raw" returned. You can do that by telling ODPI-C you want a variable with oracleTypeNum=DPI_ORACLE_TYPE_LONG_VARCHAR or DPI_ORACLE_TYPE_LONG_RAW and nativeTypeNum=DPI_NATIVE_TYPE_BYTES. ODPI-C will handle all of the dynamic fetching and allocation that is required.
***********************************************

So How do we set the oracleTypeNum, nativeTypeNum in my example so the blob column s returned as a []byte? Is it even possible in my scenario?

**Reply by tgulacsi** [reply]
I've tried it, but it says "ORA-00000: DPI-1014: conversion between Oracle type 2019 and native type 3004 is not implemented"
so no, the shortcut is not implemented here.

BTW I'd be cautious to use it here: you can SELECT several objects, and if you retrieve all the LOB attributes by default, you can drive yourself into a very slow and memory-consuming corner...

The data DOES have to be retrieved from the DB with OCI calls - that's what ODPI does as a shortcut in some situations.
But the data has to travel through the net, and you can decrease the n umber of round-trips by wrapping the lobReader into a bufio.NewReaderSize with a big enough buffer, or use io.Copy, that uses the freshly implemented lobReader.WriteTo.

**Reply by tgulacsi** [reply]
TL;DR

For big LOBs you want this streaming behaviour, as you don't want to slurp all the data into memory.

For small LOBs, if you roughly know your LOB's size, you can limit the number of round-trips to retrieve the LOB's data by using ioutil.ReadAll(bufio.NewReaderSize(lob, 4<<20)) (4Mb buffer).

**Reply by ntubati** [reply]
Thank You.. I am using ioutil.ReadAll in my scenario. But I didn't mention godror.LobAsReader() in ExecContext.. which brings me another question. What is the significance of  godror.LobAsReader() ? when we always get the reference of BLOB and CLOB objects?

**Reply by tgulacsi** [reply]
As ODPI does not provide this convenience functionality for LOB attributes of objects, It could only be done in Go.
Which is way easier than on the C side, so for now I'll just document it.

**Comment by tgulacsi** [comment]
PTAL TestIssue134 in z_test.go at 0e001d6 !

**Reply by ntubati** [reply]
"message": "getObjectType(\"XXCTS_EEM_INVC_DOC_TAB\") conn=0x7f39c807af10: ORA-00000: DPI-1062: unexpected OCI return value 1041 in function dpiConn_getObjectType"

I am getting the above error sometimes and when I restart my service, it is working fine.. could you please help?

**Reply by tgulacsi** [reply]
What version of godror are you using?
What version of Oracle Client?
What version of Oracle DB are you connecting to?

I don't see what id this 1041 means, and the ODPI library raises this error, so you should ask them (github.com/oracle/odpi).

**Reply by tgulacsi** [reply]
You should set DPI_DEBUG_LEVEL=28 env var and log the output, to see what does ODPI call before this error is raised.

**Reply by ntubati** [reply]
> What version of godror are you using? 
      v0.16.1
> What version of Oracle Client?
     oracle-instant client 19.6
> What version of Oracle DB are you connecting to?
    12.1.0.2.0

Here is are the steps in the code. 

// Get the Table type object
filterType, err = godror.GetObjectType(ctx, db.client, "XXCTS_EEM_INVC_DOC_TAB")

// Prepare Context
stmt, err := db.client.PrepareContext(ctx, "BEGIN APPS.XXCTS_EEM_TPL_INVCDOCS_PUB.RETRIEVE_SHIPPING_DOCS( :p_cco_id_i,:p_partner_name_i,:p_tp_reference_num_i,:p_doc_information_tbl_o,:p_ret_code_o,:p_ret_message_o); END;")

// Execute Context
_, err = stmt.ExecContext(ctx,
		sql.Named("p_cco_id_i", lccoID),
		sql.Named("p_partner_name_i", lpartnerName),
		sql.Named("p_tp_reference_num_i", ltprefNum),
		sql.Named("p_doc_information_tbl_o", sql.Out{Dest: Ofbject}),
		sql.Named("p_ret_code_o", sql.Out{Dest: &code}),
		sql.Named("p_ret_message_o", sql.Out{Dest: &message}),
	)


Also could you please let me know how and where to  set DPI_DEBUG_LEVEL=28  ?

**Reply by sudarshan12s** [reply]
export DPI_DEBUG_LEVEL=28 on the shell where you are running the program. It should output all the ODPI calls made to the terminal output.

**Reply by ntubati** [reply]
Thanks , i can see the log messages from ODPI call now in the terminal. .will try to find more information when the error occurs again

**Comment by ntubati** [comment]
Looks like The issue is happening when the DB is down for maintenance and our application which is in openshift container is up.
When we restart the pod the issue is gone. Is it expected behavior?

**Reply by tgulacsi** [reply]
It is expected to have lots of strange errors when the db is down, but you use the same connection.

**Reply by tgulacsi** [reply]
I've added some code (a3aa30f) to close the connection for this error.

**Reply by ntubati** [reply]
Hi , 

I am currently using v0.24.2. I see that now a change in behavior 
after the DB Is down and up , i am getting the error "getObjectType(\"XXCTS_EEM_INVC_DOC_TAB\") conn=0x7f39c807af10: ORA-00000: DPI-1062: unexpected OCI return value 1041 in function dpiConn_getObjectType" for the first time.
When I try again the error is gone without restarting the container. Previously I had to restart to get rid of the error.

Is this because of the fix you did?

**Reply by tgulacsi** [reply]
No. It's in  a3aa30fb482183fcb03fdc3fa87a80b7416fb8fd which is later than v0.24.2 - not released yet.

**Comment by sudarshan12s** [comment, marked as answer]
just to clarify on 1041 error, yes it corresponds to ORA-1041 and is
expected when connection to DB is lost.
The way currently its reported has an issue and shall be corrected in
coming OCI releases.

Thanks
